### PR TITLE
Feature manage bookings IN #31

### LINKED ISSUE #31: IN: Manage bookings
Blocked by: Manage bookings UI

Integrate management of bookings with requests so that it has the ability to 
1. accept/decline a booking
2. Depending on availability, differentiate between upcoming bookings and ones you still have to manage

To test incoming requests, by this point you should have created controllers for requests, you can create multiple requests to make sure it's integrated well

### PR DESCRIPTION
**Pull request by MarianaES**
### What this PR does (required):
- Implement buttons to accept or decline a request.
- Calendar shows booked dates (from start to end) in the past, present, and future. 
- Handle the messages to show the user when no requests exist.

### Screenshots / Videos (front-end only):
https://user-images.githubusercontent.com/29843870/138170382-e36bc6b5-940d-422a-a44a-b2944a2de454.mov

image

### Any information needed to test this feature (required):
- It is necessary to have a populated DB, specifically a Requests collection.

### CODE REVIEW

File: client/src/helpers/APICalls/requests.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import { User } from '../../interface/User';
+
+export async function getRequestList(): Promise<RequestsList> {
+  const fetchOptions: FetchOptions = {
+    method: 'GET',
+    credentials: 'include',
+  };
+
+  return await fetch('/request/', fetchOptions).then((res) =>
+    res.json().catch(() => ({ error: { message: 'Unable to connect to server. Please try again' } })),
```
Comment (ezeilo-su) [inline client/src/helpers/APICalls/requests.ts:12]: You are not chaining `catch(() =>` the correct place. You need to chain it like this: `.then(...).catch(...)`. As long as each `then` executes a `return` statement, all Promise rejections will be caught in one place. But here you're only `catch`ing the rejection from `json()` and not that from `fetch()`

File: client/src/helpers/APICalls/requests.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+): Promise<Request> {
+  const fetchOptions: FetchOptions = {
+    method: 'POST',
+    headers: { 'Content-Type': 'application/json' },
+    body: JSON.stringify({ sitterId, startDate, endDate, serviceType, totalPrice }),
+    credentials: 'include',
+  };
+  return await fetch('/request/', fetchOptions).then((res) =>
+    res.json().catch(() => ({ error: { message: 'Unable to connect to server. Please try again' } })),
+  );
```
Comment (ezeilo-su) [inline client/src/helpers/APICalls/requests.ts:31]: The comment as above

File: client/src/helpers/APICalls/requests.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+): Promise<{ request: Request }> {
+  const fetchOptions: FetchOptions = {
+    method: 'PUT',
+    headers: { 'Content-Type': 'application/json' },
+    body: JSON.stringify({ status, startDate, endDate, serviceType, totalPrice }),
+    credentials: 'include',
+  };
+  return await fetch(`/request/${requestId}`, fetchOptions).then((res) =>
+    res.json().catch(() => ({ error: { message: 'Unable to connect to server. Please try again' } })),
+  );
```
Comment (ezeilo-su) [inline client/src/helpers/APICalls/requests.ts:50]: The same promise chaining style with this one too - :smiley:

File: client/src/helpers/APICalls/requests.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+export async function updateRequest(
+  requestId: string,
+  status: string,
+  startDate?: Date,
+  endDate?: Date,
+  serviceType?: string,
+  totalPrice?: number,
+): Promise<{ request: Request }> {
+  const fetchOptions: FetchOptions = {
+    method: 'PUT',
```
Comment (ezeilo-su) [inline client/src/helpers/APICalls/requests.ts:43]: Consider using a `PATCH` request if you're not to update the whole resource.

**Review by sundayezeilo — CHANGES_REQUESTED** [review]
Great progress on this one :+1:  There are just a few comments to look at.
Happy coding!

**Review by rajivtitus — COMMENTED** [review]
Great work with this! 🚀 

Just a couple of comments from Sunday need to be address. Thanks!

File: client/src/helpers/APICalls/requests.ts (hunk 4 above)
Comment (MarianaES) [reply client/src/helpers/APICalls/requests.ts]: Yes, thanks. Same recommendation on the Request controller PR, so I changed it to PUT instead.

File: client/src/pages/Booking/Booking.tsx [hunk 5]
```diff
@@ -45,9 +53,46 @@ const materialTheme = createMuiTheme({
...
+    async function getRequestsAndSave() {
+      const allRequests = await getRequestList();
+      setRequests(allRequests);
+    }
+
+    getRequestsAndSave();
+  }, []);
+
+  const requestsReceived = requests?.requestsReceived;
+  console.log({ requestsReceived });
```
Comment (rajivtitus) [inline client/src/pages/Booking/Booking.tsx]: Please remove

File: client/src/pages/Booking/Booking.tsx [hunk 6]
```diff
@@ -74,7 +132,23 @@ function Booking(): JSX.Element {
...
+                    from={request.startDate}
+                    to={request.endDate}
+                    status={request.status}
+                    upcoming
+                  />
+                ))
+              ) : (
+                <Alert severity="info">
+                  <AlertTitle>Info</AlertTitle>
+                  You will not have pet visits in the near future 🔮 — <strong>BE PATIENT!</strong>
```
Comment (rajivtitus) [inline client/src/pages/Booking/Booking.tsx]: Please avoid using CAPITALCASE text, use a CSS class to transform text instead :)

File: client/src/pages/Booking/Booking.tsx [hunk 7]
```diff
@@ -100,11 +173,41 @@ function Booking(): JSX.Element {
...
+                    requestId={request._id}
+                    from={request.startDate}
+                    to={request.endDate}
+                    status={request.status}
+                  />
+                ))
+              ) : (
+                <Alert severity="info">
+                  <AlertTitle>Info</AlertTitle>
+                  You are not taking care of any pets for now. 🛀 — <strong>RELAX!</strong>
```
Comment (rajivtitus) [inline client/src/pages/Booking/Booking.tsx]: Same as comment above!

File: client/src/pages/Booking/Booking.tsx [hunk 8]
```diff
@@ -100,11 +173,41 @@ function Booking(): JSX.Element {
...
+                    requestId={request._id}
+                    from={request.startDate}
+                    to={request.endDate}
+                    status={request.status}
+                  />
+                ))
+              ) : (
+                <Alert severity="info">
+                  <AlertTitle>Info</AlertTitle>
+                  Do not worry, you will be able to see your memories 🐾 — <strong>SOON!</strong>
```
Comment (rajivtitus) [inline client/src/pages/Booking/Booking.tsx]: Ditto here!

File: client/src/pages/Booking/Booking.tsx (hunk 6 above)
Comment (rajivtitus) [inline client/src/pages/Booking/Booking.tsx]: Really nice touch adding Unicode emojis 🚀

**Review by rajivtitus — APPROVED** [review]
Great work with this 😄 

Just a couple of minor comments to address before merging. Thank you!

**Pull request merged**